### PR TITLE
abi encoding/decoding refactor, fixed #277

### LINKED ISSUE #277: Wrong return parameter decoding BigNumber instead of Array of BigNumbers
I have a problem with abi decoding of return arrays in web3. The following is a part of the ABI:
```
{
"inputs": [{"type": "uint256[]", "name": "event_description_hashes"}], 
"constant": false, "type": "function", "name": "getMarkets",
 "outputs": [{"type": "uint256[1024]", "name": ""}]
}
```

For some reason I don't get an array back but a single large integer in web3:

```
interface_contract.getMarkets.call(event_hashes, 'pending', function (err, markets) { 
    console.log(markets) // is  1.0151422069548598394975132446711695856e+38
});
```

### PR DESCRIPTION
**Pull request by debris**
- abi encoding/decoding refactor
- fixed #277

not done yet:
- nested dynamic array encoding. eg:
`int[][]`

improvements for future:
- reduce number of regex matches

### CODE REVIEW

File: lib/solidity/address.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+var f = require('./formatters');
+var SolidityType = require('./type');
+
+/**
+ * SolidityTypeAddress is a prootype that represents address type
```
Comment (frozeman) [inline lib/solidity/address.js]: type "prootype"

File: lib/solidity/address.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+var SolidityTypeAddress = function () {
+    this._inputFormatter = f.formatInputInt;
+    this._outputFormatter = f.formatOutputAddress;
+};
+
+SolidityTypeAddress.prototype = new SolidityType({});
+SolidityTypeAddress.prototype.constructor = SolidityTypeAddress;
+
+SolidityTypeAddress.prototype.isType = function (name) {
+    return !!name.match(/address(\[([0-9]*)\])?/);
```
Comment (frozeman) [inline lib/solidity/address.js]: why not using utils.isAddress here?

File: lib/solidity/bytes.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
+var f = require('./formatters');
+var SolidityType = require('./type');
+
+/**
+ * SolidityTypeBytes is a prootype that represents bytes type
```
Comment (frozeman) [inline lib/solidity/bytes.js]: Typo 'prootype' here too (i'll leave others)

**Comment by frozeman** [comment]
Looks nicely modularised. i can't say much about its working as i haven't been to deep into value en/decoding, but the tests pass, so thats a good sign.

there are some errors in travis though:
```
lib/solidity/coder.js: line 107, col 6, Missing semicolon.
lib/solidity/coder.js: line 115, col 19, 'e' is already defined.
lib/solidity/coder.js: line 123, col 16, 'i' is already defined.
lib/solidity/coder.js: line 125, col 19, 'e' is already defined.
lib/solidity/coder.js: line 102, col 58, This function's cyclomatic complexity is too high. (5)
lib/solidity/coder.js: line 149, col 20, 'i' is already defined.
lib/solidity/coder.js: line 157, col 24, 'nestedName' is already defined.
lib/solidity/coder.js: line 158, col 36, 'nestedStaticPartLength' is already defined.
lib/solidity/coder.js: line 159, col 20, 'result' is already defined.
lib/solidity/coder.js: line 161, col 28, 'previousLength' is already defined.
lib/solidity/coder.js: line 163, col 24, 'i' is already defined.
lib/solidity/coder.js: line 170, col 20, 'i' is already defined.
lib/solidity/coder.js: line 171, col 34, 'additionalOffset' is already defined.
lib/solidity/coder.js: line 133, col 53, This function's cyclomatic complexity is too high. (13)
lib/solidity/coder.js: line 214, col 7, Missing semicolon.
lib/solidity/coder.js: line 23, col 5, 'BigNumber' is defined but never used.
lib/solidity/coder.js: line 24, col 5, 'utils' is defined but never used.
lib/solidity/coder.js: line 25, col 5, 'SolidityParam' is defined but never used.
lib/solidity/coder.js: line 28, col 5, 'SolidityType' is defined but never used.
lib/solidity/dynamicbytes.js: line 44, col 62, 'name' is defined but never used.
lib/solidity/string.js: line 41, col 56, 'name' is defined but never used.
lib/solidity/type.js: line 89, col 20, 'length' is already defined.
lib/solidity/type.js: line 90, col 24, 'nestedName' is already defined.
lib/solidity/type.js: line 92, col 20, 'result' is already defined.
lib/solidity/type.js: line 128, col 20, 'length' is already defined.
lib/solidity/type.js: line 129, col 24, 'arrayStart' is already defined.
lib/solidity/type.js: line 131, col 24, 'nestedName' is already defined.
lib/solidity/type.js: line 132, col 36, 'nestedStaticPartLength' is already defined.
lib/solidity/type.js: line 133, col 20, 'result' is already defined.
lib/solidity/type.js: line 135, col 20, 'i' is already defined.
lib/solidity/type.js: line 142, col 20, 'length' is already defined.
lib/solidity/type.js: line 148, col 16, 'length' is already defined.
lib/solidity/type.js: line 112, col 42, This function's cyclomatic complexity is too high. (6)
lib/solidity/type.js: line 19, col 43, 'name' is defined but never used.
lib/solidity/type.js: line 30, col 53, 'name' is defined but never used.
lib/solidity/type.js: line 44, col 51, 'name' is defined but never used.
lib/solidity/type.js: line 58, col 50, 'name' is defined but never used.
lib/solidity/type.js: line 62, col 50, 'name' is defined but never used.
lib/utils/utils.js: line 111, col 31, 'escape' is not defined.
lib/utils/utils.js: line 122, col 11, 'unescape' is not defined.
```

File: lib/solidity/address.js (hunk 2 above)
Comment (debris) [inline lib/solidity/address.js]: this regexp has nothing todo with `utils.isAddress`

**Pull request merged**